lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -19,35 +17,50 @@ def initialize(serializer, options = {})
         def serializable_hash(options = nil)
           options ||= {}
           if serializer.respond_to?(:each)
-            serializer.each do |s|
-              result = self.class.new(s, @options.merge(fieldset: @fieldset)).serializable_hash(options)
-              @hash[:data] << result[:data]
-
-              if result[:included]
-                @hash[:included] ||= []
-                @hash[:included] |= result[:included]
-              end
-            end
-
-            add_links(options)
+            serializable_hash_for_collection(serializer, options)
           else
-            primary_data = primary_data_for(serializer, options)
-            relationships = relationships_for(serializer)
-            included = included_for(serializer)
-            @hash[:data] = primary_data
-            @hash[:data][:relationships] = relationships if relationships.any?
-            @hash[:included] = included if included.any?
+            serializable_hash_for_single_resource(serializer, options)
           end
-          @hash
         end
 
         def fragment_cache(cached_hash, non_cached_hash)
           root = false if @options.include?(:include)
-          ActiveModel::Serializer::Adapter::JsonApi::FragmentCache.new().fragment_cache(root, cached_hash, non_cached_hash)
+          ActiveModel::Serializer::Adapter::JsonApi::FragmentCache.new.fragment_cache(root, cached_hash, non_cached_hash)
         end
 
         private
 
+        def serializable_hash_for_collection(serializer, options)
+          hash = { data: [] }
+          serializer.each do |s|
+            result = self.class.new(s, @options.merge(fieldset: @fieldset)).serializable_hash(options)
+            hash[:data] << result[:data]
+
+            if result[:included]
+              hash[:included] ||= []
+              hash[:included] |= result[:included]
+            end
+          end
+
+          if serializer.paginated?
```

Comment:
this check could be inside `links_for` method. Today we have just pagination link, but for the future other links need be included as well.

something like that
```
hash[:links]  = links_for(serializer, options)
```

```
def links_for(serializer, options)
  return {} unless serializer.paginated?
  JsonApi::PaginationLinks.new(serializer.object, options[:context]).serializable_hash(options)
end
```

Is that make sense?